File: Task/Numeric-error-propagation/F-Sharp/numeric-error-propagation.fs
let sqr (x : float) = x * x
let abs (x : float)  = System.Math.Abs x
let pow = System.Math.Pow

type Approx (value : float, sigma : float) =
    member this.value = value
    member this.sigma = sigma

    static member (~-) (x : Approx) = Approx (- x.value, x.sigma)
    static member (%+) (x: Approx, y : float) = Approx (x.value + y, x.sigma)
    static member (%+) (y : float, x : Approx) = x %+ y
    static member (%+) (x : Approx, y : Approx) =
        Approx (x.value + y.value, sqrt((sqr x.sigma)+(sqr y.sigma)))
    static member (%-) (x: Approx, y : float) = Approx (x.value - y, x.sigma)
    static member (%-) (y : float, x : Approx) = (-x) %+ y
    static member (%-) (x : Approx, y : Approx) = x %+ (-y)
    static member (%*) (x : Approx, y : float) = Approx (y * x.value, abs(y * x.sigma))
    static member (%*) (y : float, x : Approx) = x %* y
    static member (%*) (x : Approx, y : Approx) =
        let v = x.value * y.value
        Approx (v, v * sqrt(sqr(x.sigma/x.value))+sqr(y.sigma/y.value))
    static member (%/) (x : Approx, y : Approx) =
        Approx (x.value / y.value, sqrt(sqr(x.sigma/x.value))+sqr(y.sigma/y.value))
    static member (%^) (x : Approx, y : float) =
        if y < 0. then failwith ("Cannot raise the power with a negative number " + y.ToString())
        let v = pow(x.value,y)
        Approx (v, abs(v * y * x.sigma / x.value))

    override this.ToString() = sprintf "%.2f ±%.2f" value sigma

[<EntryPoint>]
let main argv =
    let x1 = Approx (100., 1.1)
    let y1 = Approx (50., 1.2)
    let x2 = Approx (200., 2.2)
    let y2 = Approx (100., 2.3)

    printfn "Distance: %A" ((((x1 %- x2) %^ 2.) %+ ((y1 %- y2) %^ 2.)) %^ 0.5)
    0

File: Task/Numeric-error-propagation/PascalABC.NET/numeric-error-propagation.pas
type
  Approx = record
  private
    value, sigma: real;
  public
    constructor(v, s: real);
    begin
      value := v;
      sigma := s;
    end;

    class function operator-(a: Approx) := new Approx(-a.value, a.sigma);
    class function operator+(a: Approx) := a;

    class function operator+(a, b: Approx) := new Approx(a.value + b.value, (a.sigma.Sqr + b.sigma.Sqr).Sqrt);
    class function operator+(a: Approx; b: real) := new Approx(a.value + b, a.sigma);
    class function operator+(b: real; a: Approx) := a + b;

    class function operator-(a, b: Approx) := a + (-b);
    class function operator-(a: Approx; b: real) := a + (-b);
    class function operator-(b: real; a: Approx) := (-a) + b;

    class function operator*(a, b: Approx): Approx;
    begin
      var v := a.value * b.value;
      Result := new Approx(v, v * ((a.sigma / a.value).Sqr + (b.sigma / b.value).Sqr).Sqrt)
    end;

    class function operator*(a: Approx; b: real) := new Approx(a.value * b, abs(a.sigma * b));
    class function operator*(b: real; a: Approx) := a * b;

    class function operator/(a, b: Approx): Approx;
    begin
      var v := a.value / b.value;
      Result := new Approx(v, v * ((a.sigma / a.value).Sqr + (b.sigma / b.value).Sqr).Sqrt);
    end;

    class function operator/(a: Approx; b: real) := new Approx(a.value / b, abs(a.sigma * b));
    class function operator/(b: real; a: Approx) := a / b;

    class function operator**(a: Approx; b: real): Approx;
    begin
      var v := power(a.value, b);
      Result := new Approx(v, abs(v * b * a.sigma / a.value));
    end;

    function ToString: string; override;
    begin
      Result := Format('{0} ± {1}', value, sigma);
    end;
  end;

function Sqrt(Self: Approx): Approx; extensionmethod;
begin
  result := Self ** 0.5;
end;

begin
  var x1 := new Approx(100, 1.1);
  var y1 := new Approx(50, 1.2);
  var x2 := new Approx(200, 2.2);
  var y2 := new Approx(100, 2.3);
  println(((x1 - x2) ** 2 + (y1 - y2) ** 2).Sqrt);
end.
